### PR TITLE
Add `verify` feature

### PR DESCRIPTION
**Pull request by prxgr4mm3r**

### CODE REVIEW

File: src/commands/contract/compile.ts [hunk 1]
```diff
@@ -16,6 +16,12 @@ export class CompileContract extends SwankyCommand<typeof CompileContract> {
       description:
         "A production contract should always be build in `release` mode for building optimized wasm",
     }),
+    verifiable: Flags.boolean({
+      default: false,
+      char: "v",
+      description:
+        "A production contract should be build in `verifiable` mode to deploy on a public network",
```
Comment (ipapandinas) [inline src/commands/contract/compile.ts:23]: Maybe we can extend the description by mentioning: 'Ensure Docker Engine is up and running.'

**Review by ipapandinas — COMMENTED** [review]
When building locally a flipper template contract, I am getting this error:

`$ ../bin/run.js contract compile flipper --verifiable`

```bash
Compiling flipper v0.1.0 (/tmp/cargo-contract_CoFkgq/contracts/flipper)
   Compiling metadata-gen v0.1.0 (/tmp/cargo-contract_CoFkgq/contracts/flipper/.ink/metadata_gen)
    Finished release [optimized] target(s) in 10m 47s
     Running `/contract/target/ink/flipper/release/metadata-gen`
 [==] Generating bundle
ERROR: Error decoding BuildResult:
 }
Caused by:
    expected value at line 1 column 1
Error Compiling flipper contract
    at ChildProcess.<anonymous> (file://~/Documents/dev/astar/swanky-cli/dist/commands/contract/compile.js:65:36)
```

I have the same dependencies as the Dockerimage:
- cargo-contract: v4.0.0-rc.1
- rust & toolchain: 1.72


When using the devcontainer I am also facing an error

```bash
root ➜ /workspaces/temp_project (main) $ ../bin/run.js contract compile flipper --verifiable
✖ Error Compiling flipper contract
error:   ProcessError: [==] Image does not exist. Pulling one from the registry
4.0.0-rc.1: Pulling from paritytech/contracts-verifiable 
1f3513f9501d: Pull complete
d2f5b4d64af6: Pull complete
50c38c4856e5: Pull complete
0bf705fe9e85: Pull complete
Digest: sha256:463b910e18793dd1f8f8e1c92ce72f8937892f8b657525573f47ae59545703e1
Status: Downloaded newer image for paritytech/contracts-verifiable:4.0.0-rc.1
ERROR: Docker responded with status code 500: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: waiting for init preliminary setup: read init-p: connection reset by peer: unknown
Error Compiling flipper contract
    at ChildProcess.<anonymous> (file:///workspaces/dist/commands/contract/compile.js:65:36)
```

Does it works well on your side @prxgr4mm3r ? How do you proceed to test it?

**Comment by ipapandinas** [comment]
I am also wondering if these feature must be released once parity supports the `--verifiable` argument in a stable release, since the current one, 4.0.0-rc.1, is a release candidate @vsofiya

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -1,7 +1,11 @@
 {
   "name": "swanky-env",
-  "image": "ghcr.io/swankyhub/swanky-cli/swanky-base:swanky3.1.0-beta.0_v2.1.0",
-
+  "image": "ghcr.io/swankyhub/swanky-cli/swanky-base:swanky3.1.0-beta.0_v2.1.1",
```
[the remark this answers is not in the document]
Comment (vsofiya) [reply .devcontainer/devcontainer.json:3]: @prxgr4mm3r please note that swankyhub org was redone into inkdevhub and use the right links

**Comment by prxgr4mm3r** [comment]
@ipapandinas It works localy on my machine. I test it this way:
1. I create new swanky project
`➜  swanky-cli git:(feature/verifiable) ✗ ../swanky-cli/bin/run.js init ../temp_proj
? Which contract template should we use initially? flipper
? What should we name your initial contract? flipper
? What is your name? prxgr4mm3r
? What is your email? 
? Do you want to download Swanky node? Yes
✔ Copying common template files OK
✔ Checking dependencies OK
✔ Copying contract template files OK
✔ Processing templates OK
✔ Installing dependencies OK
✔ Initializing git OK
✔ Downloading Swanky node OK
✔ Writing config OK
🎉 😎 Swanky project successfully initialised! 😎 🎉

➜  swanky-cli git:(feature/verifiable) ✗ cd ../temp_proj 
`
2. From temp_project foulder run
`➜  temp_proj git:(master) ✗ ../swanky-cli/bin/run.js contract compile flipper --verifiable

The contract was built in RELEASE mode.

Your contract artifacts are ready. You can find them in:
/home/prxgr4mm3r/RustroverProjects/temp_proj/target/ink/flipper

  - flipper.contract (code + metadata)
  - flipper.wasm (the contract's code)
  - flipper.json (the contract's metadata)
✔ flipper Contract compiled successfully
✔ Moving artifacts OK
✔ flipper contract's TS types Generated successfully
`
3. Verify compiled contract

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -1,7 +1,11 @@
...
+  "features": {
+    "ghcr.io/devcontainers/features/docker-in-docker:2": {
```
[the remark this answers is not in the document]
Comment (prxgr4mm3r) [reply .devcontainer/devcontainer.json:5]: Yes, exactly.

File: src/commands/check/index.ts [hunk 4]
```diff
@@ -96,14 +100,16 @@ export default class Check extends SwankyCommand<typeof Check> {
           const mismatched: Record<string, string> = {};
           Object.entries(ctx.versions.contracts).forEach(([contract, inkPackages]) => {
             Object.entries(inkPackages).forEach(([inkPackage, version]) => {
-              if (semver.gt(version, supportedInk!)) {
-                mismatched[
-                  `${contract}-${inkPackage}`
-                ] = `Version of ${inkPackage} (${version}) in ${contract} is higher than supported ink version (${supportedInk})`;
-              }
+              if (inkPackage != "verified") {
```
Comment (ipapandinas) [inline src/commands/check/index.ts:103]: We can only search for the ink dependency version here:

```typescript
{
  title: "Verify ink version",
  task: async (ctx) => {
    const supportedInk = ctx.swankyConfig?.node.supportedInk;

    const mismatched: Record<string, string> = {};
    Object.entries(ctx.versions.contracts).forEach(([contract, contractData]) => {
      if (Object.prototype.hasOwnProperty.call(contractData, "ink")) {
        const version = contractData.ink;
        if (version && semver.gt(version, supportedInk!)) {
          mismatched[
            `${contract}-ink`
          ] = `Version of ink (${version}) in ${contract} is higher than supported ink version (${supportedInk})`;
        }

        if (!(version.startsWith("=") || version.startsWith("v"))) {
          ctx.looseDefinitionDetected = true;
        }
      }
    });

    ctx.mismatchedVersions = mismatched;
  },
},
```

File: .devcontainer/devcontainer.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (ipapandinas) [reply .devcontainer/devcontainer.json:5]: Cool, is the verification working inside your devcontainer?

File: src/lib/tasks.ts [hunk 5]
```diff
@@ -136,3 +137,27 @@ export async function installDeps(projectPath: string) {
     await execaCommand(installCommand, { cwd: projectPath });
   }
 }
+
+export function checkCargoVersion(minimalVersion: string, invalidVersionsList: string[]) {
+  const regex = /cargo-contract-contract (.*)-unknown-x86_64-unknown-linux-gnu/;
```
Comment (ipapandinas) [inline src/lib/tasks.ts:142]: This regex do not work on Mac where `cargo-contract --version` output looks like this:

`cargo-contract 4.0.0-rc.1-unknown-aarch64-apple-darwin`

Make it compatible for all architecture

File: src/lib/tasks.ts [hunk 6]
```diff
@@ -136,3 +137,27 @@ export async function installDeps(projectPath: string) {
...
+  } catch {
+    cargoVersion = null;
+  }
+  if (cargoVersion) {
+    const match = cargoVersion.match(regex);
+    if (match) {
+      cargoVersion = match[1];
+    }
+  } else {
+    throw new InputError("Verifiable mode requires cargo-contract version >= 4.0.0-rc");
```
Comment (ipapandinas) [inline src/lib/tasks.ts:156]: Do not hardcode the version here, use `minimalVersion` instead.
Same below.

**Review by ipapandinas — COMMENTED** [review]
Last adjustments

**Review by pmikolajczyk41 — APPROVED** [review]
I run the experiment locally and everything seems to be working well (both positive verification (similar steps as already presented in the discussion), and the negative one, when I changed contract code after verifiable build)

File: src/lib/tasks.ts [hunk 7]
```diff
@@ -136,3 +137,27 @@ export async function installDeps(projectPath: string) {
     await execaCommand(installCommand, { cwd: projectPath });
   }
 }
+
+export function checkCargoVersion(minimalVersion: string, invalidVersionsList: string[]) {
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/lib/tasks.ts:141]: exactly, there is an overlap here that will be resolve on merging!

File: .devcontainer/devcontainer.json [hunk 8]
```diff
@@ -1,7 +1,11 @@
 {
   "name": "swanky-env",
-  "image": "ghcr.io/swankyhub/swanky-cli/swanky-base:swanky3.1.0-beta.0_v2.1.0",
-
+  "image": "ghcr.io/inkdevhub/swanky-cli/swanky-base:swanky3.1.0-beta.0_v2.1.1",
+  "features": {
+    "ghcr.io/devcontainers/features/docker-in-docker:2": {
+      "version" : "latest"
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply .devcontainer/devcontainer.json:6]: yes, you're right!

File: src/commands/contract/compile.ts [hunk 9]
```diff
@@ -16,6 +16,12 @@ export class CompileContract extends SwankyCommand<typeof CompileContract> {
       description:
         "A production contract should always be build in `release` mode for building optimized wasm",
     }),
+    verifiable: Flags.boolean({
+      default: false,
+      char: "v",
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/commands/contract/compile.ts:21]: indeed, we can remove the char value for verifiable builds and only keep `--verifiable`

File: src/commands/contract/verify.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+      description: "Name of the contract to verify",
+    }),
+  };
+
+  async run(): Promise<void> {
+    const { args, flags } = await this.parse(VerifyContract);
+
+    checkCargoVersion("4.0.0", ["4.0.0-alpha"]);
+
+    if (args.contractName === undefined && !flags.all) {
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/commands/contract/verify.ts:34]: yes, the `--all` flag is a bit redundant, we can remove it

File: src/commands/contract/verify.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+      const contractPath = path.resolve("contracts", contractInfo.name);
+      this.logger.info(`"Looking for contract ${contractInfo.name} in path: [${contractPath}]`);
+      if (!(await pathExists(contractPath))) {
+        throw new InputError(`Contract folder not found at expected path`);
+      }
+
+      if(!contractInfo.build) {
+        throw new InputError(`Contract ${contractName} is not compiled. Please compile it first`);
+      }
+
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/commands/contract/verify.ts:61]: make sense 🙌

File: src/commands/contract/compile.ts [hunk 12]
```diff
@@ -65,11 +71,15 @@ export class CompileContract extends SwankyCommand<typeof CompileContract> {
...
               "--manifest-path",
-              `${contractPath}/Cargo.toml`,
+              `contracts/${contractName}/Cargo.toml`,
             ];
-            if (flags.release) {
+            if (flags.release && !flags.verifiable) {
               compileArgs.push("--release");
             }
+            if (flags.verifiable) {
+              checkCargoVersion("4.0.0", ["4.0.0-alpha"]);
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/commands/contract/compile.ts:80]: This can be fix by using `semver.satisfies` with `<4.0.0` in `checkCargoVersion` similar to #195 @prxgr4mm3r

**Pull request merged**